### PR TITLE
feat: FFXIV Strategy Board encoder/decoder

### PR DESCRIPTION
**Pull request by caasi**
## Summary

- Add a TypeScript `src/lib/ffxiv-strat/` module that decodes and encodes FFXIV strategy board `[stgy:...]` share codes
- Multi-layer codec: substitution cipher → custom base64 → CRC32 + zlib → SoA binary format
- Demo page at `/playground/ffxiv-strat` with reactive auto-decode (paste and go)
- 58 tests covering unit, integration, round-trip identity, and edge cases
- TypeScript build support added (webpack 4 + SWC + Jest 24)

## Details

**Codec layers** (`src/lib/ffxiv-strat/`):
- `types.ts` — BoardData/BoardObject interfaces, error classes, constants
- `tables.ts` — KEY_TABLE, ALPHABET_TABLE, base64 char mapping
- `codec.ts` — substitution cipher encrypt/decrypt, custom base64 encode/decode, stgy shell wrap/unwrap
- `compression.ts` — CRC32 calculation, zlib compress/decompress via pako
- `binary.ts` — SoA binary parser/serializer (BinaryReader/BinaryWriter)
- `index.ts` — public `decode()` / `encode()` API

**Build infra**:
- `tsconfig.json`, `jest-ts-transform.js` for TypeScript support
- `jest-node-resolver.js` for `node:` protocol compatibility with Jest 24
- `jest.setup.js` reordered to polyfill TextEncoder/TextDecoder before enzyme
- `cheerio` pinned to `1.0.0-rc.12` to avoid undici/Jest 24 conflicts
- `swc-loader` pinned to `0.1.15` for webpack 4 compatibility

**Attribution**: Based on [marimelon/stgy-tools](https://github.com/marimelon/stgy-tools) (MIT). See `src/lib/ffxiv-strat/LICENSE` and `NOTICE`.

## Test plan

- [x] `yarn test` — 58 tests pass (tables, codec, compression, binary, integration)
- [x] Round-trip identity: `decode(encode(decode(fixture))) ≡ decode(fixture)` for 4 real codes
- [x] Table bijectivity: KEY_TABLE and ALPHABET_TABLE have no duplicate values
- [x] MAX_OBJECTS (50) encode without u16 overflow
- [x] Error cases: malformed input, truncated payload, corrupted CRC
- [ ] `yarn build` — production build
- [ ] Visual check: `/playground/ffxiv-strat` in light and dark mode

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/lib/ffxiv-strat/index.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+  const binaryData = serializeBoardData(board);
+
+  // 2. Compress
+  const compressedData = compress(binaryData);
+
+  // 3. Build integrity header
+  const lengthBytes = new Uint8Array(2);
+  lengthBytes[0] = binaryData.length & 0xff;
+  lengthBytes[1] = (binaryData.length >> 8) & 0xff;
+
```
[the remark this answers is not in the document]
Comment (caasi) [reply src/lib/ffxiv-strat/index.ts:102]: Fixed in 921c524 — added u16 bounds check, throws StratEncodeError if binaryData exceeds 65535 bytes.

File: src/lib/ffxiv-strat/binary.ts [hunk 2]
```diff
@@ -0,0 +1,673 @@
...
+ * Serialize all field containers (TypeContainers) inside the content section.
+ */
+function serializeFields(board: BoardData, writer: BinaryWriter): void {
+  const objects = board.objects
+  const objectCount = objects.length
+
+  // Field 1: Board name
+  if (board.name.length > 0) {
+    const nameBytes = new TextEncoder().encode(board.name)
+    const nameLength = nameBytes.length
```
[the remark this answers is not in the document]
Comment (caasi) [reply src/lib/ffxiv-strat/binary.ts:616]: Fixed in a8be3d0 — added validation for MAX_OBJECTS (50), MAX_NAME_LENGTH (20), and MAX_TEXT_BYTES (30) in serializeBoardData.

File: src/lib/ffxiv-strat/binary.ts [hunk 3]
```diff
@@ -0,0 +1,673 @@
...
+    // share the same value. parseFieldTextOrBackground handles both cases
+    // by inspecting the length value (1 => background section).
+    const parser = fieldParsers[sectionTypeOrFieldId]
+    if (parser) {
+      parser(reader, context)
+    } else {
+      // Unknown field ID encountered. The binary format has no generic length
+      // prefix, so we cannot safely skip the field's data. Break out of the
+      // parsing loop to avoid misinterpreting payload bytes as field IDs.
+      break
```
[the remark this answers is not in the document]
Comment (caasi) [reply src/lib/ffxiv-strat/binary.ts]: Fixed in d4f1376 — now throws StratDecodeError on unknown field ID instead of silently breaking.

File: src/lib/ffxiv-strat/binary.ts [hunk 4]
```diff
@@ -0,0 +1,673 @@
...
+      this.view.byteOffset + this.offset,
+      length,
+    )
+    this.offset += length
+    let end = bytes.indexOf(0)
+    if (end === -1) end = bytes.length
+    return new TextDecoder('utf-8').decode(bytes.subarray(0, end))
+  }
+
+  skip(bytes: number): void {
```
[the remark this answers is not in the document]
Comment (caasi) [reply src/lib/ffxiv-strat/binary.ts:88]: Fixed in 517124f — added bytes >= 0 check and ensureRemaining() call before advancing offset.

File: jest-node-resolver.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+      fs.mkdirSync(shimDir, { recursive: true })
+    }
+    fs.writeFileSync(shimPath, `module.exports = require('${moduleName}');\n`)
+  }
+  return shimPath
+}
+
+module.exports = (request, options) => {
+  if (request.startsWith('node:')) {
+    return ensureShim(request.slice(5))
```
[the remark this answers is not in the document]
Comment (caasi) [reply jest-node-resolver.js]: Fixed in 3a41a4e — simplified to strip node: prefix and delegate to defaultResolver directly, no more filesystem side effects.

File: jest.setup.js [hunk 6]
```diff
@@ -1,6 +1,16 @@
-import 'raf/polyfill'
+// Polyfill TextEncoder/TextDecoder before any imports that might need them.
+// (undici, loaded transitively via cheerio → enzyme, requires TextDecoder)
+//
+// This file uses require() instead of import because ES module imports are
+// hoisted — the polyfill assignment must execute before enzyme's transitive
+// deps trigger side-effects that reference TextDecoder.
+const { TextEncoder, TextDecoder } = require('util');
+global.TextEncoder = TextEncoder;
+global.TextDecoder = TextDecoder;
```
[the remark this answers is not in the document]
Comment (caasi) [reply jest.setup.js]: Fixed in e684347 — guarded with null checks so native TextEncoder/TextDecoder won't be overridden.

File: src/lib/ffxiv-strat/codec.ts [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+ * Encodes binary data to a URL-safe base64 string (no padding).
+ *
+ * @param data - Raw bytes to encode.
+ * @returns URL-safe base64 string (`-` and `_` instead of `+` and `/`).
+ */
+export function encodeBase64(data: Uint8Array): string {
+  let binary = '';
+  for (let i = 0; i < data.length; i++) {
+    binary += String.fromCharCode(data[i]);
+  }
```
[the remark this answers is not in the document]
Comment (caasi) [reply src/lib/ffxiv-strat/codec.ts]: Fixed in ddf308c — switched to chunked String.fromCharCode(...subarray) with 0x8000 chunk size.

File: src/lib/ffxiv-strat/index.ts [hunk 8]
```diff
@@ -0,0 +1,106 @@
+import { StratDecodeError, StratEncodeError } from './types';
+import type { BoardData } from './types';
+import { KEY_TABLE, REVERSE_KEY_TABLE, base64CharToValue } from './tables';
+import { unwrapStgy, wrapStgy, decryptCipher, encryptCipher, encodeBase64, decodeBase64 } from './codec';
+import { calculateCRC32, packHeader, unpackHeader, compress, decompress } from './compression';
+import { parseBoardData, serializeBoardData } from './binary';
```
[the remark this answers is not in the document]
Comment (caasi) [reply src/lib/ffxiv-strat/index.ts]: Fixed in 4aa2957 + 0682c6c — removed trailing semicolons and aligned formatting. Also upgraded prettier from 1.16.4 to 3.8.2 so it can parse TypeScript import type syntax.

File: docs/superpowers/specs/2026-04-12-ffxiv-strat-codec-design.md [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+
+**Header (16 bytes):**
+
+| Offset | Size | Type    | Description                                         |
+|--------|------|---------|-----------------------------------------------------|
+| 0x00   | 4    | u32 LE  | Version (expected: 2). Reject other values.         |
+| 0x04   | 4    | u32 LE  | Content length (bytes after header)                 |
+| 0x08   | 2    | u16 LE  | Flags (observed: `0x0001`). Preserve for round-trip.|
+| 0x0A   | 2    | u16 LE  | Title length (bytes)                                |
+| 0x0C   | 4    | —       | Padding (zeros). Preserve for round-trip.           |
```
[the remark this answers is not in the document]
Comment (caasi) [reply docs/superpowers/specs/2026-04-12-ffxiv-strat-codec-design.md]: Fixed in a90abae — updated spec and code. Research shows flags=0x0001 is a constant present in all real game data (wtw0212 documents it as 'Flag | Fixed 01 00'). Encoder now hardcodes 0x0001 instead of 0. Spec updated to reflect this.

File: src/lib/ffxiv-strat/codec.ts [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+export function decodeBase64(base64: string): Uint8Array {
+  let std = base64.replace(/-/g, '+').replace(/_/g, '/')
+  const pad = (4 - (std.length % 4)) % 4
+  std += '='.repeat(pad)
+  const binary = atob(std)
+  const bytes = new Uint8Array(binary.length)
+  for (let i = 0; i < binary.length; i++) {
+    bytes[i] = binary.charCodeAt(i)
+  }
+  return bytes
```
[the remark this answers is not in the document]
Comment (caasi) [reply src/lib/ffxiv-strat/codec.ts:99]: Fixed in bb4fcff — wrapped atob() in try/catch, rethrows as StratDecodeError.

File: src/lib/ffxiv-strat/binary.ts [hunk 11]
```diff
@@ -0,0 +1,702 @@
...
+    // by inspecting the length value (1 => background section).
+    const parser = fieldParsers[sectionTypeOrFieldId]
+    if (parser) {
+      parser(reader, context)
+    } else {
+      // Unknown field ID encountered. The binary format has no generic length
+      // prefix, so we cannot safely skip the field's data. Treat this as a
+      // decode failure instead of returning a partially parsed board.
+      throw new StratDecodeError(
+        `Unknown field ID: 0x${sectionTypeOrFieldId.toString(16).padStart(4, '0')}`
```
[the remark this answers is not in the document]
Comment (caasi) [reply src/lib/ffxiv-strat/binary.ts:478]: Updated spec in bea11d7 — Field 0x09 never appears in real data (game client skips it). The parser throws StratDecodeError on unknown fields because the binary format has no generic length prefix for safe skipping. Spec updated to match implementation.

File: src/lib/ffxiv-strat/binary.ts [hunk 12]
```diff
@@ -0,0 +1,702 @@
...
+
+  // Field 5: Positions array (raw 1/10 pixel values, no scaling)
+  if (objectCount > 0) {
+    writer.writeUint16(FieldIds.POSITIONS)
+    writer.writeUint16(3) // type = 3 (two u16 per element)
+    writer.writeUint16(objectCount)
+    for (const obj of objects) {
+      writer.writeUint16(obj.position.x)
+      writer.writeUint16(obj.position.y)
+    }
```
[the remark this answers is not in the document]
Comment (caasi) [reply src/lib/ffxiv-strat/binary.ts:672]: Fixed in bcf2ff7 — added u16/u8/i16 range validation for all numeric fields (objectId, position, rotation, size, color, params, backgroundId). Throws StratEncodeError on out-of-range values.

File: package.json [hunk 13]
```diff
@@ -58,7 +59,7 @@
     "postcss-loader": "3",
     "postcss-scss": "2",
     "precss": "4",
-    "prettier": "1.16.4",
+    "prettier": "3",
```
[the remark this answers is not in the document]
Comment (caasi) [reply package.json]: Fixed in 862afa1 — pinned to exact version 3.8.2.

File: package.json [hunk 14]
```diff
@@ -20,7 +20,7 @@
     "deploy": "gh-pages -t -d dist",
     "test": "jest",
     "codecov": "jest && codecov",
-    "prettier": "npx prettier --single-quote --no-semi --trailing-comma es5 --arrow-parens always --write '*.js' 'config/**/*.js'",
+    "prettier": "npx prettier --single-quote --no-semi --trailing-comma es5 --arrow-parens always --write '*.js' 'config/**/*.js' 'src/**/*.ts'",
```
[the remark this answers is not in the document]
Comment (caasi) [reply package.json]: Fixed in 9432f77 — added src/**/*.tsx to prettier glob.

File: jest.setup.js [hunk 15]
```diff
@@ -1,6 +1,20 @@
...
+// This file uses require() instead of import because ES module imports are
+// hoisted — the polyfill assignment must execute before enzyme's transitive
+// deps trigger side-effects that reference TextDecoder.
+const { TextEncoder, TextDecoder } = require('util');
+if (global.TextEncoder == null) {
+  global.TextEncoder = TextEncoder;
+}
+if (global.TextDecoder == null) {
+  global.TextDecoder = TextDecoder;
+}
```
[the remark this answers is not in the document]
Comment (caasi) [reply jest.setup.js:13]: Fixed in 9432f77 — ran prettier on jest.setup.js, semicolons removed.

File: jest-ts-transform.js [hunk 16]
```diff
@@ -0,0 +1,29 @@
+/**
+ * Custom Jest transformer for TypeScript files using @swc/core directly.
+ * Needed because @swc/jest v0.2.2 caches its config globally on first use,
+ * making it incompatible with per-extension config splits in Jest 24.
+ */
+const { transformSync } = require('@swc/core');
+
+module.exports = {
```
[the remark this answers is not in the document]
Comment (caasi) [reply jest-ts-transform.js:8]: Fixed in 9432f77 — jest-ts-transform.js was already formatted correctly (no semicolons, single quotes). Confirmed by running prettier.
